### PR TITLE
Allow for setting the jujushell URL in the local storage.

### PR DESCRIPTION
**Pull request by frankban**
The GUI settings (shift+!) can be used to set a jujushell dns name, either as a remote http(s) or ws(s) address.

### CODE REVIEW

**Comment by frankban** [comment]
QA: see description.

**Comment by jujugui** [comment]
Refer to this link for build results (access rights to CI server needed): 
http://ci-gce.jujugui.org:8080//job/juju-gui/8349/

File: jujugui/static/gui/src/app/components/modal-gui-settings/modal-gui-settings.js [hunk 1]
```diff
@@ -108,6 +135,7 @@ class ModalGUISettings extends React.Component {
 
 ModalGUISettings.propTypes = {
   closeModal: PropTypes.func.isRequired,
+  flags: PropTypes.object.isRequired,
```
Comment (mitechie) [inline jujugui/static/gui/src/app/components/modal-gui-settings/modal-gui-settings.js]: was this previously missing? I don't see that it was added in the code above and not sure what flags is here atm.

File: jujugui/static/gui/src/app/init/component-renderers-mixin.js [hunk 2]
```diff
@@ -430,6 +423,7 @@ Browser: ${navigator.userAgent}`
     ReactDOM.render(
       <ModalGUISettings
         closeModal={this._clearSettingsModal.bind(this)}
+        flags={this.applicationConfig.flags}
```
Comment (mitechie) [inline jujugui/static/gui/src/app/init/component-renderers-mixin.js]: ah ok, here's the flags bit.

File: jujugui/static/gui/src/app/init/component-renderers-mixin.js [hunk 3]
```diff
@@ -232,14 +232,7 @@ Browser: ${navigator.userAgent}`
...
-    const address = function() {
-      if (db.environment.get('jujushellURL')) {
-        return `ws://${db.environment.get('jujushellURL')}/ws/`;
-      }
-      if (config.jujushellURL) {
-        return config.jujushellURL;
-      }
-    }();
+    const address = initUtils.jujushellURL(localStorage, db, config);
     if (!address) {
```
Comment (mitechie) [inline jujugui/static/gui/src/app/init/component-renderers-mixin.js]: should this not be more explicit in this error? If we don't have an address we know what's up and the generic "unknown error" seems confusing?

File: jujugui/static/gui/src/app/init/test-utils.js [hunk 4]
```diff
@@ -99,6 +99,55 @@ describe('init utils', () => {
...
+    it('returns the URL stored in the storage', () => {
+      assertURL({storage: 'wss://1.2.3.4/ws1/'}, 'wss://1.2.3.4/ws1/');
+      assertURL({storage: 'ws://1.2.3.4/ws2/'}, 'ws://1.2.3.4/ws2/');
+      assertURL({storage: '1.2.3.4/ws3/'}, 'wss://1.2.3.4/ws3/');
+      assertURL({storage: 'http://1.2.3.4/ws4/'}, 'ws://1.2.3.4/ws4/');
+      assertURL({storage: 'https://1.2.3.4/ws5/'}, 'wss://1.2.3.4/ws5/');
+    });
+
+    it('returns the URL stored in the db', () => {
+      assertURL({db: '1.2.3.4'}, 'ws://1.2.3.4/ws/');
```
Comment (mitechie) [inline jujugui/static/gui/src/app/init/test-utils.js]: based on this test, I was expecting my value in the input box (10.0.0.1) to turn into wss://10.0.0.1/ws, however checking the failed wss connection (only shows in chrome dev tools) it tried to connect to wss://10.0.0.1/

File: jujugui/static/gui/src/app/init/component-renderers-mixin.js (hunk 3 above)
Comment (frankban) [reply jujugui/static/gui/src/app/init/component-renderers-mixin.js]: I made this clearer.

File: jujugui/static/gui/src/app/init/test-utils.js (hunk 4 above)
Comment (frankban) [reply jujugui/static/gui/src/app/init/test-utils.js]: This is a test for when the address is obrained from the mega-watcher. In which case, we ad /ws/ as we can assume the path. In the other cases, users are free to set the path manually and we don't force /ws/.

Comment (frankban) [reply jujugui/static/gui/src/app/init/test-utils.js]: Let's chat about this.

**Comment by jujugui** [comment]
Refer to this link for build results (access rights to CI server needed): 
http://ci-gce.jujugui.org:8080//job/juju-gui/8354/

**Comment by frankban** [comment]
Thanks for the review Rick!
:shipit:

**Comment by jujugui** [comment]
Status: merge request accepted. Url: http://ci-gce.jujugui.org:8080/job/juju-gui-merge

**Pull request merged**